### ISSUE TITLE
handling circular dependencies

### ISSUE THREAD

**Issue by tnrich** [erikras/ducks-modular-redux#24]
Hey there,
First off, thanks for the great work you've been doing (redux-form in particular has made my life much easier) 

I'm running into a an issue right now that I think is related to circular dependencies between two modules which both require action types from one another. 

I'm styling my code in the "ducks" fashion, where I'm exporting a default reducer, as well as my action creators, but I'm also using a third-party library, 'redux-act'. Here's what my code looks like: 

####Module 1: Which relies on one of module 2's exported action creator
```js
//./selectionLayer.js
import { updateCaret } from './caretPosition';
import { createAction, createReducer } from 'redux-act'

// ------------------------------------
// Actions
// ------------------------------------
export const clearSelection = createAction('clearSelection')
export const updateSelectionLayer = createAction('updateSelectionLayer')

// ------------------------------------
// Reducer
// ------------------------------------
export default createReducer({
  [updateCaret]: () => {
    //clear the selection layer if the caret is updated!
    return {
      start: -1,
      end: -1
    }
  },
  [clearSelection]: () => {
    return {
      start: -1,
      end: -1
    }
  },
  [updateSelectionLayer]: (state, newSelectionLayer) => {
    if (process.env.NODE_ENV !== 'production') {
      if (!(newSelectionLayer.start >= 0 && newSelectionLayer.end >= 0)) {
        throw new Error('we should never be here! updateSelectionLayer must always be called with a valid selection layer')
      }
    }
    return newSelectionLayer
  },
}, {
  start: -1,
  end: -1
})
```

####Module 2: Which relies on one of module 1's exported action creator
```js
//./caretPosition.js
import { updateSelectionLayer } from './selectionLayer';
import { createAction, createReducer } from 'redux-act'

// ------------------------------------
// Actions
// ------------------------------------
export const clearCaret = createAction('clearCaret')
export const updateCaret = createAction('updateCaret')

// ------------------------------------
// Reducer
// ------------------------------------
export default createReducer({
  [updateSelectionLayer]: () => {
  	//clear the caret if the selection layer is updated!
    return -1
  },
  [clearCaret]: () => {
    return -1
  },
  [updateCaret]: ({start}, payload) => {
    return payload
  },
}, 0)

```

Basically the issue is the updateCaret action creator is undefined within my `./selectionLayer.js` file, but if I comment out the circular dependency, it becomes defined again. 

I thought I'd found an answer to my issue: http://stackoverflow.com/questions/30378226/circular-imports-with-webpack-returning-empty-object, but I've tried putting my import statements lower down in the files, but to no avail. I'll ask about this issue in the redux-act library as well. 

Maybe this is something to do with es6's circular dependency resolution? 

Any help would be greatly appreciated. Thanks!
Thomas

**Comment by erikras** [comment]
Beats me. This might be a failure point of the ducks paradigm.

You've got yourself a duck circle.

![ducks](https://cloud.githubusercontent.com/assets/4396759/13255144/fc503d06-da44-11e5-93f4-26651e6700bf.jpg)

Drawing by @blesh. https://github.com/reactjs/redux/issues/151#issuecomment-137833403

**Comment by tnrich** [comment]
Hmm thanks for the reply and the nice picture lol. Is this a known issue? Have other people run into this before? 

If it truly is a failure point within the ducks system, how would recommend avoiding/fixing it? Should I make a separate action creators file just for this one case? Or mush the two reducers into one file?

Thanks for the help!
Thomas

**Comment by tnrich** [comment]
Hmm.. so I tried using commonJS require syntax, instead of es6's "import" syntax, and now it seems to work: 

```js
//./caretPosition.js
import { createAction, createReducer } from 'redux-act'

// ------------------------------------
// Actions
// ------------------------------------
export const clearCaret = createAction('clearCaret')
export const updateCaret = createAction('updateCaret')


var { updateSelectionLayer } = require('./selectionLayer'); //important, keep this here in the order of the file!
window.updateSelectionLayer = updateSelectionLayer
debugger;
// ------------------------------------
// Reducer
// ------------------------------------
export default createReducer({
  [updateSelectionLayer]: () => {
  	//clear the caret if the selection layer is updated!
    return -1
  },
  [clearCaret]: () => {
    return -1
  },
  [updateCaret]: ({start}, payload) => {
    return payload
  },
}, 0)

```

```
//./selectionLayer.js
import { createAction, createReducer } from 'redux-act'

// ------------------------------------
// Actions
// ------------------------------------
export const clearSelection = createAction('clearSelection')
export const updateSelectionLayer = createAction('updateSelectionLayer')

var { updateCaret } = require('./caretPosition'); //important, keep this here in the order of the file!
window.updateCaret = updateCaret
debugger;
// ------------------------------------
// Reducer
// ------------------------------------
export default createReducer({
  [updateCaret]: () => {
    //clear the selection layer if the caret is updated!
    return {
      start: -1,
      end: -1
    }
  },
  [clearSelection]: () => {
    return {
      start: -1,
      end: -1
    }
  },
  [updateSelectionLayer]: (state, newSelectionLayer) => {
    if (process.env.NODE_ENV !== 'production') {
      if (!(newSelectionLayer.start >= 0 && newSelectionLayer.end >= 0)) {
        throw new Error('we should never be here! updateSelectionLayer must always be called with a valid selection layer')
      }
    }
    return newSelectionLayer
  },
}, {
  start: -1,
  end: -1
})

```

**Comment by pauldijou** [comment]
You can always extract common action creators into a separated file which will be imported by both reducers.

**Comment by blesh** [comment]
LOL... I love that I've been summoned to this issue by my duck drawing.

**Comment by erikras** [comment]
@blesh :smile: 

@pauldijou has the right answer here, I think.

**Comment by wmertens** [comment]
I think the webpack behavior is a bug, no? ES6 clearly allows circular imports (as long as the cycle is finite)
=> https://github.com/webpack/webpack/issues/1788

**Comment by deoxxa** [comment]
The problem here is that you're using the imported values before they're ready. Here's a small example:

a.js

```js
import { B } from './b';

export const A = 'A';

console.log('immediate in a.js:', B);

setTimeout(() => console.log('timeout in a.js:', B), 1);
```

b.js:

```js
import { A } from './a';

export const B = 'B';

console.log('immediate in b.js:', A);

setTimeout(() => console.log('timeout in b.js:', A), 1);
```

When you run `a.js` you'll see something like this:

```
immediate in b.js: undefined
immediate in a.js: B
timeout in b.js: A
timeout in a.js: B
```

When you run `b.js` you'll see something like this:

```
immediate in a.js: undefined
immediate in b.js: A
timeout in a.js: B
timeout in b.js: A
```

When you do this:

```js
export default createReducer({
  [updateCaret]: () => { return ...; },
}, ...);
```

you're accessing `updateCaret` before the module where it comes from has finished being evaluated. This is because you're accessing it at module evaluation time. If you can push that back to a point in time after the whole module is evaluated, webpack/babel/browser-from-the-future will have a chance to fill in all the circular references.

My solution to this (since I had a very similar problem stemming from a very similar structure) was to move to something more like this:

```ks
export default (state = ..., action) => {
  switch (action.type) {
  case updateCaret: { return ...; }
  }
};
```

I don't love it, because it's not as nice to read, but it did the job.

**Comment by joshwcomeau** [comment]
I'm finding myself in a similar predicament, where Duck A needs to dispatch an action from Duck B. I didn't want to import Duck B's action creators, because that will lead to circular dependencies. 

I think a good solution (at least for my use-case) is something like [Redux Saga](https://github.com/yelouafi/redux-saga).

Essentially, you'd have a Saga that imports all the action creators it needs (from Duck A and Duck B). Then, you'd dispatch an action that the saga would listen to, and orchestrate from there. 

```js
// random_thing.saga.js

import { actionA, TRIGGER_ACTION } from 'duckA';
import { actionB } from 'duckB';

function* watchTriggerAction() {
  while (true) {
    const action = yield take(TRIGGER_ACTION);
    
    yield put(actionA());
    yield put(actionB());
  }
}
```

**Comment by wmertens** [comment]
Would this be solvable by using the `System.import().then()`? https://github.com/ModuleLoader/es6-module-loader/blob/v0.17.0/docs/circular-references-bindings.md

**Comment by mrchess** [comment]
@tnrich So did you have success just switching over to `require` and using webpack? I am running into the same issue. Am about to try rewriting some stuff using `require` instead of import.

**Comment by alvelig** [comment]
I think it depends on what you want to achieve. `import` does dependency on build time, when `require` does it on runtime, when your circular dependency is already resolved. Thus it gives a valid solution. But the right solution is to resolve your circular dependency moving to a separate file.

**Comment by jabacchetta** [comment]
I just ran into this issue with selectors. Good post that explains the problem: http://www.datchley.name/scoped-selectors-for-redux-modules/
